File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.py
from collections import Counter


def solution(gems):
    n = len(gems)
    k = len(set(gems))
    
    left, right = 0, k - 2
    answer = [1, n]
    count = Counter(gems[left:right + 1])
    
    while right < n - 1:
        right += 1
        count[gems[right]] += 1
        while len(count) == k:
            if right - left < answer[1] - answer[0]:
                answer[0], answer[1] = left + 1, right + 1
            count[gems[left]] -= 1
            if count[gems[left]] == 0:
                count.pop(gems[left])
            left += 1

    return answer

File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.java
import java.util.*;

class Solution {
    public int[] solution(String[] gems) {
        HashSet<String> gemType = new HashSet<>();
        HashMap<String, Integer> counter = new HashMap<>();
        
        for (String gem : gems) {
            gemType.add(gem);
        }
        
        int types = gemType.size();
        int left = 0, right = 0;
        counter.put(gems[0], 1);
        int[] answer = {0, 100_000};
        
        while (true) {
            while (counter.size() == types) {
                if (right - left < answer[1] - answer[0]) {
                    answer[0] = left + 1;
                    answer[1] = right + 1;
                }
                
                counter.put(gems[left], counter.get(gems[left]) - 1);
                
                if (counter.get(gems[left]) == 0) {
                    counter.remove(gems[left]);
                }
                
                left++;
            }
            
            if (counter.size() < types) {
                if (right + 1 == gems.length)
                    break;
                counter.put(gems[++right], counter.getOrDefault(gems[right], 0) + 1);
            }
        }
        
        return answer;
    }
}
